File: files/vskel.ts
/*
struct Foo {
  x: int?,
  bar: [[Bar]]
}

struct Bar {
  y: int
}
*/

/*
{
  "x": null,
  "bar": [
    [
      {
        "y": 71
      },
      {
        "y": 72
      }
    ],
    [
      {
        "y": 41
      },
      {
        "y": 42
      }
    ]
  ]
}
*/

export type Constructor<T> = new (...arguments_: any) => T;
export type Instance<I> = I extends Constructor<infer T> ? T : never;

let _messageMap: Map<string, Constructor<StructMessage>> = new Map();
let _fieldsMap: Map<Constructor<StructMessage>, StructField[]> = new Map();

abstract class StructMessage {}

const enum TyKindTag {
  Primitive,
  Message,
  Array,
}

type MessageClassID = string | Constructor<StructMessage>;

type TyKind =
  | { kind: TyKindTag.Primitive }
  | { kind: TyKindTag.Message; of: MessageClassID }
  | { kind: TyKindTag.Array; of: TyKind };

type StructField = {
  name: string;
  ty: TyKind;
};

function valueToPlainObject(value: any, ty: TyKind) {
  if (value === null) {
    return null;
  }
  //
  else if (ty.kind === TyKindTag.Primitive) {
    return value;
  }
  //
  else if (ty.kind === TyKindTag.Message) {
    let result: Record<string, any> = {};

    let fields = _fieldsMap.get(Object.getPrototypeOf(value).constructor)!;
    for (let i = 0; i < fields.length; ++i) {
      const { name: fieldName, ty } = fields[i];
      const fieldVal = value[fieldName];
      result[fieldName] = valueToPlainObject(fieldVal, ty);
    }

    return result;
  }
  //
  else if (ty.kind === TyKindTag.Array) {
    return (value as any[]).map((val) => valueToPlainObject(val, ty.of));
  } else {
    throw new Error("Invalid value/ty");
  }
}

function plainObjectToValue(obj: any, ty: TyKind) {
  if (obj === null) {
    return null;
  }
  //
  else if (ty.kind === TyKindTag.Primitive) {
    return obj;
  } else if (ty.kind === TyKindTag.Message) {
    let ctor = typeof ty.of === "string" ? _messageMap.get(ty.of)! : ty.of;
    let fields = _fieldsMap.get(ctor)!;

    let createPayload = {};
    for (let i = 0; i < fields.length; ++i) {
      const { name: fieldName, ty } = fields[i];
      const fieldVal = obj[fieldName];
      createPayload[fieldName] = plainObjectToValue(fieldVal, ty);
    }

    // TODO: Make this strongly typed
    return (ctor as any).create(createPayload);
  }
  //
  else if (ty.kind === TyKindTag.Array) {
    return (obj as any[]).map((val) => plainObjectToValue(val, ty.of));
  }
  //
  else {
    throw new Error("Invalid value/ty");
  }
}

namespace forser {
  export function packMessage<M extends StructMessage>(message: M) {
    return JSON.stringify(
      valueToPlainObject(message, {
        kind: TyKindTag.Message,
        of: Object.getPrototypeOf(message).constructor,
      })
    );
  }

  export function unpackMessage<M extends Constructor<StructMessage>>(
    messageType: M,
    serialized: string
  ): Instance<M> {
    let obj = JSON.parse(serialized);
    return plainObjectToValue(obj, {
      kind: TyKindTag.Message,
      of: messageType,
    });
  }
}

type FieldsOf<T> = Pick<
  T,
  {
    [K in keyof T]: T[K] extends Function ? never : K;
  }[keyof T]
>;

/* =========================================== */

const _FooFields: StructField[] = [
  { name: "x", ty: { kind: TyKindTag.Primitive } },
  {
    name: "bar",
    ty: {
      kind: TyKindTag.Array,
      of: { kind: TyKindTag.Array, of: { kind: TyKindTag.Message, of: "Bar" } },
    },
  },
];

export class Foo extends StructMessage {
  static create(body: FieldsOf<Foo>): Foo {
    return Object.assign(new Foo(), body);
  }

  public x!: number | null;
  public bar!: Bar[][];
}

_messageMap.set("Foo", Foo);
_fieldsMap.set(Foo, _FooFields);

/* =========================================== */

const _BarFields: StructField[] = [
  { name: "y", ty: { kind: TyKindTag.Primitive } },
];

export class Bar extends StructMessage {
  static create(body: FieldsOf<Bar>): Bar {
    return Object.assign(new Bar(), body);
  }

  public y!: number;
}

_messageMap.set("Bar", Bar);
_fieldsMap.set(Bar, _BarFields);

/* =========================================== */


let foo1 = Foo.create({
  x: null,
  bar: [
    [Bar.create({ y: 71 }), Bar.create({ y: 72 })],
    [Bar.create({ y: 41 }), Bar.create({ y: 42 })],
  ],
});

console.log(Object.getPrototypeOf(foo1.bar[0][0]).constructor);

let json = `{
  "x": null,
  "bar": [
    [
      {
        "y": 71
      },
      {
        "y": 72
      }
    ],
    [
      {
        "y": 41
      },
      {
        "y": 42
      }
    ]
  ]
}`;

let foo2 = forser.unpackMessage(Foo, json);
console.log(Object.getPrototypeOf(foo2.bar[0][0]).constructor);

File: files/vskel.py
"""
struct Foo {
  x: int?,
  bar: [[Bar]]
}

struct Bar {
  y: int
}
"""

from __future__ import annotations
from typing import Literal, Optional, cast, Type, TypeVar, Union, Dict, Any, List
from dataclasses import dataclass
import json

T = TypeVar('T', bound='StructMessage')
MessageClassID = Union[str, Type[T]]

class StructMessage:
    pass

@dataclass
class TyKind:
    kind: Literal['primitive', 'message', 'array']
    of: TyKind | None = None
    ctor: MessageClassID | None = None

@dataclass
class StructField:
    name: str
    kind: TyKind

_fields_map: Dict[Type[StructMessage], List[StructField]] = {}

def _value_to_plain_object(value: Any, ty: TyKind) -> Any:
    if value is None:
        return None
    
    elif ty.kind == 'primitive':
        return value

    elif ty.kind == 'message':
        result: Dict[str, Any] = {}

        fields = _fields_map[value.__class__]
        for f in fields:
            result[f.name] = _value_to_plain_object(
                getattr(value, f.name),
                f.kind
            )

        return result

    elif ty.kind == 'array':
        arr = cast(list[Any], value)
        inner = cast(TyKind, ty.of)
        return list(map(
            lambda val: _value_to_plain_object(val, inner),
            arr
        ))
    else:
        raise ValueError("Invalid value/ty")

def pack_message(message: StructMessage):
    return json.dumps(
        _value_to_plain_object(
            message,
            TyKind('message', ctor=message.__class__)
        )
    )

# ================================ #

_FooFields: list[StructField] = [
    StructField('x', kind=TyKind('primitive')),
    StructField('bar', kind=TyKind(
        'array',
        of=TyKind(
            'array',
            of=TyKind(
                'message',
                ctor='Bar'
            )
        )
    )),
]

@dataclass
class Foo(StructMessage):
    x: Optional[int]
    bar: list[list['Bar']]

_fields_map[Foo] = _FooFields

# ================================ #

_BarFields: list[StructField] = [
    StructField('y', kind=TyKind('primitive')),
]

@dataclass
class Bar(StructMessage):
    y: int

_fields_map[Bar] = _BarFields 

foo = Foo(
    x=45,
    bar=[
        [Bar(y=1), Bar(y=2)],
        [Bar(y=3), Bar(y=4)],
    ]
)

print(pack_message(foo))